### ISSUE TITLE
False-positive in prefer-const and array-destructuring

### ISSUE THREAD

**Issue by danez** [eslint/eslint#8308]
**Tell us about your environment**

* **ESLint Version:** 3.18.0
* **Node Version:** 6.10.0
* **npm Version:** 4.4.4

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```json
{
  "ecmaVersion": 7,
  "rules": [
    "prefer-const"
  ]
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
let predicate;
[typeNode.returnType, predicate] = foo();
```

**What did you expect to happen?**

No error

**What actually happened? Please include the actual, raw output from ESLint.**

```
2:23 - 'predicate' is never reassigned. Use 'const' instead. (prefer-const)
```

The rule should consider that it is not possible to actually make predicate a const here.

**Comment by not-an-aardvark** [comment]
Thanks for the report. I was able to reproduce this issue.

**Comment by VictorHom** [comment]
I would like to look into this issue as well if no one is working on this :)

**Comment by not-an-aardvark** [comment]
Thanks @VictorHom!

**Comment by pvdlg** [comment]
If that can help, the same happen with Object destructuring:
```js
let a;
const b = {};

({a, c: b.c} = func());
```
`4:3 - 'a' is never reassigned. Use 'const' instead. (prefer-const)`

**Comment by nzakas** [comment]
This seems like a pretty significant bug. I'd like to take a stab at fixing this.
